lib/services/task_list_service.dart

```diff
@@ -23,141 +27,340 @@ class TaskListService with LogMixin, ChangeCallbackProvider {
     this._syncService,
   );
 
-  Future<List<Task>> get tasks async {
-    return (await _taskListCrdt).values;
-  }
+  Future<Iterable<TaskRecord>> get allTaskRecords async =>
+      _decodeTasks(await _readAndDecodeTaskListCollectionCrdt()).map(
+        (v) => TaskRecord(
+          v.taskRecord.value,
+          v.taskRecord.clock.node,
+          DateTime.fromMillisecondsSinceEpoch(v.taskRecord.clock.timestamp),
+          v.taskListId,
+        ),
+      );
 
-  Future<List<Task>> getTasksForList(TaskList taskList) async {
-    var sorted = <Task>[];
+  Future<Iterable<Task>> get allTasks async =>
+      (await taskLists).map((taskList) => taskList.elements).expand((e) => e);
 
-    List allTasks = await tasks;
-    for (var i = 0; i < allTasks.length; i++) {
-      if (allTasks[i].taskListID == taskList.id) {
-        sorted.add(allTasks[i]);
-      }
-    }
+  Future<Iterable<TaskList>> get taskLists async => (await _taskListRecords)
+      .where((record) => !record.isDeleted)
+      .map((record) => record.value!);
+
+  Future<TaskList?> getTaskListById(String taskListId) async =>
+      (await _taskListRecordMap)[taskListId]?.value;
+
+  Future<Iterable<Task>> getTasksFromList(String taskListId) async =>
+      getTaskListById(taskListId)
+          .then((taskList) => taskList != null ? taskList.elements : []);
+
+  Future<Iterable<Record<TaskList>>> get _taskListRecords async =>
+      (await _taskListRecordMap).values;
 
-    switch (taskList.sortBy) {
-      case SortOption.Title:
-        sorted.sort((a, b) => a.title.toString().compareTo(b.title.toString()));
-        break;
-      case SortOption.Flag:
-        sorted.sort((a, b) {
-          if (b.isFlagged) {
-            return 1;
-          }
-
-          return 0;
-        });
-        break;
-      case SortOption.Status:
-        sorted.sort((a, b) {
-          if (b.completed) {
-            return 0;
-          }
-
-          return 1;
-        });
-        break;
-      case SortOption.DueDate:
-        sorted.sort((a, b) {
-          if (b.due == null) {
-            return 0;
-          }
-          if (a.due == null) {
-            return 1;
-          }
-
-          return a.due!.compareTo(b.due!);
-        });
-        break;
-      case SortOption.Created:
-        break;
+  Future<Map<String, Record<TaskList>>> get _taskListRecordMap async =>
+      _decodeTaskListCollection(await _readAndDecodeTaskListCollectionCrdt());
+
+  /// Returns false if the task list does not exist
+  Future<bool> upsertTask(String taskListId, Task task) async {
+    l.info('Upsert task ${task.toJson()}');
+    final crdt = await _readAndDecodeTaskListCollectionCrdt();
+    final taskListCrdt = crdt.get(taskListId);
+    if (taskListCrdt == null) return false;
+
+    task.id ??= Uuid().v4();
+    final taskNode = taskListCrdt.getRecord(task.id!);
+    if (taskNode == null || taskNode.isDeleted) {
```

Comment:
If this does not become to excessive, could you add a docstring explaining what a taskNode is / how the CRDT is structured? Currently, I don't understand why taskNode == null and its deletion is handled the same. (Obviously I don't understand this, because I don't know the CRDT functionality in detail. So if it is possible to summarize this in a docstring/comment would be great!)